src/net.rs

```diff
@@ -1,25 +1,179 @@
+    let now = time();
+    let mut resolved_addrs = match lookup_host_with_timeout(hostname, port, timeout_val).await {
+        Ok(res) => res,
+        Err(err) => {
+            warn!(
+                context,
+                "DNS resolution for {}:{} failed: {:#}.", hostname, port, err
+            );
+            Vec::new()
+        }
+    };
+
+    for (i, addr) in resolved_addrs.iter().enumerate() {
+        info!(context, "Resolved {}:{} into {}.", hostname, port, &addr);
+
+        let i = i64::try_from(i).unwrap_or_default();
+
+        // Update the cache.
+        //
+        // Add sequence number to the timestamp, so addresses are ordered by timestamp in the same
+        // order as the resolver returns them.
+        context
+            .sql
+            .execute(
+                "INSERT INTO dns_cache
+                 (hostname, port, address, timestamp)
+                 VALUES (?, ?, ?, ?)
+                 ON CONFLICT (hostname, port, address)
+                 DO UPDATE SET timestamp=excluded.timestamp",
+                paramsv![
+                    hostname,
+                    port,
+                    addr.to_string(),
+                    now.saturating_add(i).saturating_add(1)
+                ],
+            )
+            .await?;
+    }
+
+    if load_cache {
+        for cached_address in context
+            .sql
+            .query_map(
+                "SELECT address
+                 FROM dns_cache
+                 WHERE hostname = ?
+                 AND ? < timestamp + 30 * 24 * 3600
+                 ORDER BY timestamp DESC",
+                paramsv![hostname, now],
+                |row| {
+                    let address: String = row.get(0)?;
+                    Ok(address)
+                },
+                |rows| {
+                    rows.collect::<std::result::Result<Vec<_>, _>>()
+                        .map_err(Into::into)
+                },
+            )
+            .await?
+        {
+            match SocketAddr::from_str(&cached_address) {
+                Ok(addr) => {
+                    if !resolved_addrs.contains(&addr) {
+                        resolved_addrs.push(addr);
+                    }
+                }
+                Err(err) => {
+                    warn!(
+                        context,
+                        "Failed to parse cached address {:?}: {:#}.", cached_address, err
+                    );
+                }
+            }
+        }
+    }
+
+    Ok(resolved_addrs)
+}
+
 /// Returns a TCP connection stream with read/write timeouts set
 /// and Nagle's algorithm disabled with `TCP_NODELAY`.
 ///
 /// `TCP_NODELAY` ensures writing to the stream always results in immediate sending of the packet
 /// to the network, which is important to reduce the latency of interactive protocols such as IMAP.
+///
+/// If `load_cache` is true, may use cached DNS results.
+/// Use this only if the connection is going to be protected with TLS.
```

Comment:
I'm wondering if this comment should extended, because passing `strict_tls` here is surprising and may even look to a reader like the wrong variable was used. Sth. like:

```rust
/// If `load_cache` is true, may use cached DNS results.
/// Use this only if the connection is going to be protected with TLS,
/// i.e. pass `false` for connecting insecurely and `strict_tls` for connections
/// that are sometimes safe, because `strict_tls=true` means that we will
/// reject invalid certificates.
```